Bug 1459430: add optional master discovery by label

wozniakjan opened the pull request:
Discovery by service name with readiness probe based on ES status code caused a deadlock. Allowing master discovery to be based on label and port allows the nodes to discover each other and readiness probe prevents them from being prematurely accessible.

The configuration file would still support service based discovery, but it could also look like:
```
cloud:
  kubernetes:
    pod_label: component=es
    pod_port: 9300
    namespace: ${NAMESPACE}
```

While currently only this is accepted:
```
cloud:
  kubernetes:
    service: ${SERVICE_DNS}
    namespace: ${NAMESPACE}
```

**Bugzilla:** https://bugzilla.redhat.com/show_bug.cgi?id=1459430

jcantrill commented on pom.xml:
```diff
@@ -20,7 +20,7 @@
   <groupId>io.fabric8</groupId>
   <artifactId>elasticsearch-cloud-kubernetes</artifactId>
-  <version>2.4.5-SNAPSHOT</version>
+  <version>2.4.4</version>
```
No need to change this as the fabric deployment process will update it correctly

jcantrill commented on src/main/java/io/fabric8/elasticsearch/cloud/kubernetes/KubernetesAPIServiceImpl.java:
```diff
@@ -28,20 +32,29 @@
+  public List<Pod> pods() {
+    logger.debug("get endpoints with pod label {}, namespace {}", podLabel, namespace);
+    final String[] l = podLabel.split("=");
+    return client().pods().inNamespace(namespace).withLabel(l[0], l[1]).list().getItems();
```
Some concerns I have here is whether it is possible to return a podlabel that is not in the pattern of 'foo=bar'?  It is possible to query for pods with labels that have a key but not a specific value.  Is the case we ALWAYS get key and value and will always be able to split it into 2 parts?  The value comes from settings?  Do we need a init check there to exit early if it is not in the expected format?

wozniakjan replied:
The pod label will be defined [here](https://github.com/openshift/openshift-ansible/blob/46d9ba942e49e4b6953dc8fa47e38d6ea98ec122/roles/openshift_logging_elasticsearch/templates/elasticsearch.yml.j2#L26). The label selector API seems to be very flexible allowing us to implement pretty much any type of selection we would like, the above algorithm was the simplest.

Should I allow selection based on the key only, or multiple labels?

jcantrill replied:
As long as we control how we get the key/value and know that's what it will be, this is sufficient for me.  We can provide further updates if we think we will need them.

jcantrill commented on src/main/java/io/fabric8/elasticsearch/discovery/kubernetes/KubernetesUnicastHostsProvider.java:
```diff
@@ -76,13 +82,81 @@ public KubernetesUnicastHostsProvider(Settings settings,
+              logger.trace("current node found. Ignoring {}", ipAddress);
+            } else {
+              endpointSubset.getPorts().stream().forEach((port) -> {
+                try {
```
Looks like there is an opportunity to refactor this block into a private method for reuse here and  the previous method

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/discovery/kubernetes/KubernetesDiscoveryPlugin.java:
```diff
@@ -53,12 +53,13 @@ public static boolean isDiscoveryAlive(Settings settings, ESLogger logger) {
 
-    if (!checkProperty(KubernetesAPIService.Fields.NAMESPACE, settings.get(KubernetesAPIService.Fields.NAMESPACE), logger) ||
-      !checkProperty(KubernetesAPIService.Fields.SERVICE_NAME, settings.get(KubernetesAPIService.Fields.SERVICE_NAME), logger)) {
+    if ( !hasNamespace(settings, logger) || !(hasServiceName(settings, logger) ^ hasPodLabel(settings, logger)) ) {
```
The '^' operater seems like a strange syntax here and is equivalent to single '|'  (I think)

wozniakjan replied:
That depends on what the desired behavior is. The `^` is an exclusive or, so it should prevent the definition of both service name and pod label. In this case, I think we should not allow defining both in the config file.

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/discovery/kubernetes/KubernetesDiscoveryPlugin.java:
```diff
@@ -67,11 +68,25 @@ public static boolean isDiscoveryAlive(Settings settings, ESLogger logger) {
       logger.warn("{} is not set.", name);
       return false;
     }
+    logger.warn("{} is set.", name);
```
Should this be a debug message?

wozniakjan commented:
> Are readiness probes in OpenShift for a service or for a particular container?

I think that is unfortunate wording because as asked, "both" as well as "neither" could be acceptable answers. A readiness probe should extend the information of the pod being temporarily unavailable to its service. That particular service is responsible for load balancing the service-to-service traffic to its pods and takes into an account status from the readiness probe. Definition of **unavailability** is left to the concrete implementation of the particular probe as it varies deeply on the application. In this case, an elastic search pod probably shouldn't be taken as ready unless its state was negotiated successfully by its internal distributed protocols and other peers recognize it fully.

> If other pods stop working, will this pods' readiness state change to be not ready?

That depends on the internal state and implementation of elastic search cluster. Elastic search itself recognizes whether the particular node is operational or not, probe picks up that information, forwards to the service, and that service handles it appropriately.
- if a minority of pods are malfunctioning but rest of the elastic search cluster is healthy and operational, then service will route requests only to the healthy majority
- if a majority of pods are malfunctioning then no pods can serve traffic and service won't be available

> How do all pods become ready if they can't be ready due to other pods not being ready?

To answer the question as worder, they never do. That is what was happening previously. This change introduces a service internal communication channel [defined exactly as the service is defined](https://kubernetes.io/docs/concepts/services-networking/service/) - by a label and a port. But bypassing mechanisms that services have implemented for service-to-service communication (load balancing and readiness checks) because they don't necessarily make sense for service internal communication (peer discovery).

portante commented:
@wozniakjan, from the description of [Services](https://kubernetes.io/docs/concepts/services-networking/service/), do we find a description of readiness probes for services?  On that page there is a link to ["Configure Liveness and Readiness Probes"](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#defining-readiness-probes), where the very first sentence reads, "This page shows how to configure liveness and readiness probes for Containers."  It goes on to describe liveness probes, "The kubelet uses liveness probes to know when to restart a Container.", and readiness probes, "The kubelet uses readiness probes to know when a Container is ready to start accepting traffic."

In order for Elasticsearch to operate as cluster, the individual containers that make up that service need to exchange message via the network.  If readiness probes are defined using information about the state of the cluster, then won't we have a chicken-and-egg problem?

How does this change solve that problem?

Does the liveness probe take into account any information about the operational state of the cluster?

Can you state and point out in the code what the liveness probe does?  What APIs it calls and what response it is expecting to get back?

Can you do the same for readiness probes?

wozniakjan commented:
> In order for Elasticsearch to operate as cluster, the individual containers that make up that service need to exchange message via the network. If readiness probes are defined using information about the state of the cluster, then won't we have a chicken-and-egg problem?

We would if we used the service for this communication (as we tried previously). But not being ready doesn't remove other pods' properties (label, port, ...), it only "removes" it from the service. The problem wasn't that nodes started initially as not ready, that is desired, the problem was they were unable to exchange messages because when contacting service it said there are no pods to communicate with.

> How does this change solve that problem?

As said above, this change introduces a service internal communication channel (defined exactly as the service is defined)[https://kubernetes.io/docs/concepts/services-networking/service/] - by a label and a port. But bypassing mechanisms that services have implemented for service-to-service communication (load balancing and readiness checks) because they don't necessarily make sense for the service internal communication (peer discovery).

> Does the liveness probe take into account any information about the operational state of the cluster?

afaik, we don't have any liveness probe for ES, because liveness probe could kill the ES pod and that could be very costly. Instead, we let ES cluster recover as it is well equipped to handle such situations.

> Can you state and point out in the code what the liveness probe does? What APIs it calls and what response it is expecting to get back?
> 
> Can you do the same for readiness probes?

https://github.com/openshift/origin-aggregated-logging/blob/master/elasticsearch/probe/readiness.sh
It calls `localhost:9200` and expects `200 OK`

wozniakjan commented:
@jcantrill I believe I addressed all comments and concerns about the label discovery for the readiness probe, do you think we can get this merged so QE can start testing?
